File: C++/H-Index.cpp
class Solution {
public:
    int hIndex(vector<int>& citations) {
        int n = citations.size();
        int count[n];
        fill_n(count, n, 0);
        for (int c : citations) {
            if (c > 0) {
                count[min(c, n)-1] ++;
            }
        }
        for (int ret = n; ret > 0; ret --) {
            if (ret < n) count[ret-1] += count[ret];
            if (count[ret-1] >= ret) return ret;
        }
        return 0;
    }
};

File: Go/H-Index.go
func hIndex(citations []int) int {
    n := len(citations)
    count_papers := make([]int, n+1)
    for i := range citations {
        if citations[i] > n {
            count_papers[n] ++
        } else {
            count_papers[citations[i]] ++
        }
    }
    if count_papers[n] >= n { return n }
    for i := n-1; i >= 0; i-- {
        count_papers[i] += count_papers[i+1]
        if (count_papers[i] >= i) { return i }
    }
    return 0
}
